doc/md/getting-started/motoko-introduction.md

```diff
@@ -1,91 +1,145 @@
+
+Key design points of Motoko include:
+
+-   Object-oriented, functional & imperative.
+
+-   Objects as records of functions.
+
+-   `async`/`await` for sequential programming of asynchronous messaging.
+
+-   Structural typing with simple generics and subtyping.
+
+-   Safe arithmetic (both unbounded and checked).
+
+-   Non-nullable types by default.
+
+-   Garbage collected (no manual memory management).
+
+-   JavaScript-like syntax but statically typed & sane.
+
+Semantics of Motoko include:
+
+-   Call-by-value (like Java, C, JS, ML; unlike Haskell).
 
-## Approachability
+-   Declarations are locally mutually recursive.
 
-Motoko is a modern language designed to be approachable for programmers who have some basic familiarity with modern object-oriented and/or functional programming idioms in either JavaScript, or another modern programming language, such as Rust, Swift, TypeScript, C#, or Java.
+-   Parametric, bounded polymorphism.
 
-## Asynchronous messaging and type sound execution
+-   Subtyping as zero-cost subsumption, not coercion.
 
-Motoko permits modern programming idioms, including special programming abstractions for distributed applications (dapps). Each dapp consists of one or more *actors* that communicate solely by *asynchronous message passing*. The state of an actor is isolated from all other actors, supporting distribution. There is no way to share state between several actors. The actor-based programming abstractions of Motoko permit human-readable message-passing patterns, and they enforce that each network interaction obeys certain rules and avoids certain common mistakes.
+-   No dynamic casts.
 
-Specifically, Motoko programs are *type sound* since Motoko includes a practical, modern type system that checks each one before it executes. The Motoko type system statically checks that each Motoko program will execute safely, without dynamic type errors, on all possible inputs. Consequently, entire classes of common programming pitfalls that are common in other languages, and web programming languages in particular, are ruled out. This includes null reference errors, mis-matched argument or result types, missing field errors and many others.
+-   No inheritance.
 
-To execute, Motoko statically compiles to [WebAssembly](about-this-guide.md#webassembly), a portable binary format that abstracts cleanly over modern computer hardware, and thus permits its execution broadly on the Internet, and the Internet Computer.
+## Implementation
 
-## Each canister smart contract as an *actor*
+Motoko is implemented in OCaml and leverages the `wasm` library. It uses a simple reference interpreter, but has a more complex compilation into WebAssembly. Compilation uses multipass with typed IR in each pass, and uniform representation, unboxed arithmetic. Motoko offers three garbage collectors (copying, compacting, or generational), and garbage collection is invoked after messages. Motoko uses polymorphism by erasure.
```

Comment:
This is irrelevant to our users, I'd say

Reply:
Removed this section